BlazarService/src/main/java/com/hubspot/blazar/listener/SlackNotificationVisitor.java

```diff
@@ -0,0 +1,199 @@
+import com.hubspot.blazar.base.externalservice.slack.SlackAttachment;
+import com.hubspot.blazar.base.externalservice.slack.SlackAttachmentField;
+import com.hubspot.blazar.base.externalservice.slack.SlackMessage;
+import com.hubspot.blazar.base.slack.SlackConfiguration;
+import com.hubspot.blazar.base.visitor.ModuleBuildVisitor;
+import com.hubspot.blazar.base.visitor.RepositoryBuildVisitor;
+import com.hubspot.blazar.data.service.BranchService;
+import com.hubspot.blazar.data.service.ModuleBuildService;
+import com.hubspot.blazar.data.service.ModuleService;
+import com.hubspot.blazar.data.service.RepositoryBuildService;
+import com.hubspot.blazar.data.service.SlackConfigurationService;
+import com.hubspot.blazar.integration.slack.SlackClient;
+import com.hubspot.blazar.util.BlazarUrlHelper;
+
+@Singleton
+public class SlackNotificationVisitor implements RepositoryBuildVisitor, ModuleBuildVisitor {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SlackNotificationVisitor.class);
+  private static final Set<RepositoryBuild.State> FAILED_REPO_STATES = ImmutableSet.of(RepositoryBuild.State.CANCELLED, RepositoryBuild.State.FAILED, RepositoryBuild.State.UNSTABLE);
+  private static final Set<ModuleBuild.State> FAILED_MODULE_STATES = ImmutableSet.of(ModuleBuild.State.CANCELLED, ModuleBuild.State.FAILED);
+  private static final Optional<String> ABSENT_STRING = Optional.absent();
+
+  private SlackConfigurationService slackConfigurationService;
+  private final BranchService branchService;
+  private ModuleService moduleService;
+  private final ModuleBuildService moduleBuildService;
+  private final BlazarUrlHelper blazarUrlHelper;
+  private final SlackClient slackClient;
+  private final RepositoryBuildService repositoryBuildService;
+
+  @Inject
+  public SlackNotificationVisitor(SlackConfigurationService slackConfigurationService,
+                                  BranchService branchService,
+                                  ModuleService moduleService,
+                                  ModuleBuildService moduleBuildService,
+                                  BlazarUrlHelper blazarUrlHelper,
+                                  SlackClient slackClient,
+                                  RepositoryBuildService repositoryBuildService) {
+    this.slackConfigurationService = slackConfigurationService;
+    this.branchService = branchService;
+    this.moduleService = moduleService;
+    this.moduleBuildService = moduleBuildService;
+    this.blazarUrlHelper = blazarUrlHelper;
+    this.slackClient = slackClient;
+    this.repositoryBuildService = repositoryBuildService;
+  }
+
+
+  @Override
+  public void visit(RepositoryBuild build) throws Exception {
+    Set<SlackConfiguration> configurationSet = slackConfigurationService.getAllWithBranchId(build.getBranchId());
+    for (SlackConfiguration slackConfiguration : configurationSet) {
+      Optional<RepositoryBuild> previous = repositoryBuildService.getPreviousBuild(build);
```

Comment:
Does the result of `repositoryBuildService.getPreviousBuild(build)` change based on the current `configurationSet` you're iterating on? If not, can you run the `previous.isPresent()` step outside of this `for` loop so that you can exit quickly?